### ISSUE TITLE
Do not create user branch until user saves the file

### ISSUE THREAD

**Issue by birchamp** [unfoldingWord/tc-create-app#888]
Currently when the user opens a file the user branch is created and the file is copied into that branch.  Users want to be able to open a file without creating a branch until the file has been saved by the user.  

On Open:   
* If the branch exists then the target side should show the contents of the branch
* If the branch does not exist then the target side should show the master, but be editable.
* If the master does not exist then the target side should show the Source, but be editable.

On Save:
* If the branch doesn’t exist then the branch should be created from the master or source (the way it is currently on open), then the changes should be saved to the branch.
* If the branch exists then the changes should be saved to the branch.

DoD:
===
Opening a file should not create a new branch until the changes have been saved.

**Comment by mandolyte** [comment]
**Notes:**
The user branch is created within the gitea-react-toolkit package (hereafter, GRT). In particular, the acquisition is attempted via the "ensure" functions. This area in GRT is written in typescript.

```ts
export const ensureContent = async ({
  config, owner, repo, branch, filepath, content, message, author, onOpenValidation,
}: ModifyContentOptions): Promise<ContentObject> => {
```

This function goes like this:
- first attempts to just read it as-if it is already there in the user branch.
- on failure, the "catch" will attempt to read it from the target repo. This will be the case if it has ever been created, updated, and merged into the master branch previsouly.
- on failure, the next "catch" will attempt to create it. Here is the code:
```ts
    } catch { // try to create the file if it doesn't exist in default or new branch
      contentObject = await createContent({
        config, owner, repo, branch, filepath, content, message, author,
      });
    };
```
- the code to create finally gets to a POST with payload to create the branch and initialize the target content with the source content.

**Comment by mandolyte** [comment]
So in theory, we should:
- avoid "updating" the branch with content when the content is found in the target master branch and, instead, just return the content object.
- if the content has never been created before (ie, not in the target master branch), we should just return a content object containing the source content.

Lastly, we should tweak the createContent function so it can be run by the "save function". Something like this:
- in the save function, first check to see if branch and file exist. If so, then proceed as normal to update the content
- otherwise, run the createContent function to create the branch and file.

**Comment by mandolyte** [comment]
I have a complete trace of the save functionality in:
https://github.com/mandolyte/uw-journal/blob/main/apps/tc-create-app/Save%20Function%20Trace.md

I'll paste a snapshot in next comment.

**Comment by mandolyte** [comment]
# Save function trace

## tc-create-app
Note: this is post-refactor. All the below are in tc-create unless otherwise noted.

*Starting with Translatable.js*

These lines:
```js
      if (filepath.match(/\.md$/)) {
        let translatableProps = {
          original: sourceFileContent,
          translation: targetFileContent,
          onTranslation: saveTranslation,
          onContentIsDirty: setContentIsDirty,
        };
        console.log('Markdown file selected');
        _translatable = <MarkdownContextProvider><MarkDownTranslatable {...translatableProps} /></MarkdownContextProvider>;
      } else if (filepath.match(/\.tsv$/)) {
        console.log('TSV file selected');
        _translatable = <TranslatableTSV onSave={saveTranslation} onEdit={autoSaveOnEdit} onContentIsDirty={setContentIsDirty} />;
```
For both markdown and tsv files, the function `saveTranslation` is passed.

*Note that auto save is there for TSV but not for markdown!*

The `saveTranslation` function comes from a hook:
```js
  const {
    actions: {
      autoSaveOnEdit,
      saveTranslation,
    },
    component: authenticationDialog,
  } = useRetrySave();
```

From the hook we find (including the auto save too):
```js
  const saveTranslation = useDeepCompareCallback(async (content) => {
    setSavingTargetFileContent(content);

    try {
      await save(content);
    } catch (error) {
      const { isRecoverable } = parseError({ error });

      // assumption is that it is an authentication issue.
      if (isRecoverable) {
        openAuthenticationDialog();
      } else {
        setSaveFailed(true);
      };
    };
  }, [save]);

  const autoSaveOnEdit = useCallback(async (content) => {
    await saveCache(content);
  }, [saveCache]);
```

The save is done by a "save" function. This, and auto save, from the AppContext:
```js
  const {
    giteaReactToolkit: {
      authenticationHook,
      targetFileHook,
    },
  } = useContext(AppContext);

  const { save, saveCache } = targetFileHook.actions || {};
```

The above hints that the functionality lies in the Gitea React Toolkit (GRT).
These lines in `App.context.js` show where it comes from:
```js
  const giteaReactToolkit = useGiteaReactToolkit({ state, actions });

  const value = {
    state,
    actions,
    giteaReactToolkit,
  };

  return <AppContext.Provider value={value}>{children}</AppContext.Provider>;
};
```

So the GRT hook is returning a `targetFileHook` which returns the save function (and the auto save cache function).

In the GRT hook code, we find:
```js
  const targetFileHook = useFile({
    config,
    authentication,
    repository: targetRepository,
    filepath: (readyForTargetFile ? filepath : undefined),
    onFilepath: setFilepath,
    defaultContent,
    onOpenValidation: _onOpenValidation,
    onLoadCache: _onLoadCache,
    onSaveCache: _onSaveCache,
  });
```

The useFile hook is in the GRT package.

## gitea-react-toolkit

Picking up with the above useFile hook
at `gitea-react-toolkit\src\components\file\useFile.js`.

The save and saveCache are part of the set of actions returned by the hook:
```js
  const actions = {
    update,
    load,
    read,
    save,
    saveCache,
    onSaveCache,
    onLoadCache,
    close,
    dangerouslyDelete,
    setIsChanged,
    onConfirmClose,
  };
```

The save function is defined in this file as:
```js
  const save = useDeepCompareCallback(async (_content) => {
    await saveFile({ authentication, repository, branch, file, content: _content });
    // (save() will not happen for "OFFLINE" system files)
    await saveCache(); // Empty cache if user has saved this file
    await load();
    contentActions.reset();
  }, [writeable, authentication, repository, branch, file, load, saveFile, saveCache]);
```

Here we see a function named `saveFile` that is handling the actual save of the content to DCS.

That function is imported:
```js
import {
  saveFile, ensureFile, deleteFile, getContentFromFile,
} from './helpers';
```

In the helpers file, we have:
```js
export const saveFile = async ({
  authentication, repository, branch, file, content, message,
}) => {
  const {
    user: author, config, token: { name: tokenid },
  } = authentication;
  const { owner: { username: owner }, name: repo } = repository;
  const { path: filepath, sha } = file;
  const _message = message || `Edit '${filepath}' using '${tokenid}'`;

  const response = await updateContent({
    config, owner, repo, branch, filepath,
    content, message: _message, author, sha,
  });
  return response;
};
```

**Note! the code assumes the file exists. Notice it uses "updateContent" unconditionally.**

Thus here is where we need to do some checking. Since we defer branch creation until this point, here is where those changes will need to be made.

There is a function in this file named "ensureFile()"; possibly it could be used with the updated content provided as the "default". Here is the function signature:

```js
export const ensureFile = async ({
  config, authentication, repository, branch, filepath, defaultContent, message, onOpenValidation,
}) => { 
```

This function uses "ensureContent" to actually do the work.

The `ensureContent()` function is here:
`src\core\gitea-api\repos\contents\contents.ts`. Note it is written in Typescript.

This function has a lot of comments to help understand what it is doing. 
Let's step thru it.

First here is the function signature and first line of code:
```js
export const ensureContent = async ({
  config, owner, repo, branch, filepath, content, message, author, onOpenValidation,
}: ModifyContentOptions): Promise<ContentObject> => {
  let contentObject: ContentObject;
```

Note that a "ContentObject" is returned. This is an interface, defined here as:
```ts
interface ContentObject {
  path: string;
  sha: string;
  content: string;
  html_url: string;
}
```

Next is the outermost try block with comments removed:
```ts
  try { 
    contentObject = await readContent({
      owner, repo, ref: branch, filepath, config,
    });
    if (!contentObject) throw new Error('File does not exist in branch');

    const _content:string = await getContentFromFile(contentObject);
    let notices: string[] = [];
    if ( onOpenValidation ) {
      notices = onOpenValidation(filepath, _content,contentObject.html_url);
    }
  } catch {
```

It first attempts to read the content from the branch. Remember that this function is trying to save the file. It is trying to ensure that the branch and file *already* exist. And, if not, then put the "default content" there.

So if there is content, that means the user has already begun to edit the file and thus it will be returned. If an "onOpenValidation" function is passed it is used to validate that the file is reasonable and won't cause the app to crash.

If the read fails, then an error is thrown. So proceed to the catch and the inner try block (with comments removed and new ones inserted)
```ts
    try { 
		
	  // does the file exist in the repo's master branch?
	  // if so, then translation work has been done before
	  // and this is an update to it. An error is thrown if none
      const _contentObject = await readContent({
        owner, repo, filepath, config,
      });

	  // this function decodes it from base64
      const _content = await getContentFromFile(_contentObject);
      let notices: string[] = [];
      if ( onOpenValidation ) {
        notices = onOpenValidation(filepath, _content,_contentObject.html_url);
      }
      if ( notices.length === 0 ) {
        // only update if no notices
        contentObject = await updateContent({
          config, owner, repo, branch, filepath, content: _content, message, author, sha: _contentObject.sha,
        });
      } else {
        contentObject = _contentObject;
      }
    } catch { // try to create the file if it doesn't exist in default or new branch
      contentObject = await createContent({
        config, owner, repo, branch, filepath, content, message, author,
      });
    };
  };
```

Here the code attempts to read from the master branch of the repo. If this file has been worked on before, it will exist there. So it is read, decoded from base64, light validation performed (optional), then the file is updated in the branch.

If the file is not in the master branch, then this is first time this file has been worked on. And an error is thrown. The to the inner catch,  which reads:

```ts
    } catch { // try to create the file if it doesn't exist in default or new branch
      contentObject = await createContent({
        config, owner, repo, branch, filepath, content, message, author,
      });
    };
```

Finally, an attempt is made create the branch (if it doesn't exist) and the file.

The "createContent" function is in this file and it uses Gitea API POSTs to do the work. The function returns a ContentObject.

The "updateContent" function is also in this file and uses PUTs to update the file.

**Comment by elsylambert** [comment]
translationCore Create - v1.6.0 build 164-985e4f7. Works for Editor mode and partially for translate mode. TC does not create user branch until Save, but in one case it does as described below:
https://qa.door43.org/translate_test/ru_tn/branches- Fails for this Repo
org- translate_test, lang- ru, resource- tn
With the above selections, branch is created on opening the file.

**Comment by superdav42** [comment]
Fix to change the behavior so files that do not exist in master will still not create a user branch until saved. https://github.com/unfoldingWord/gitea-react-toolkit/pull/141

Once this is merged we'll need to publish the gitea-react-toolkit then update tc-create-app then we can test this again

**Comment by elsylambert** [comment]
Works as expected in translationCore Create - v1.6.0 build 159-84e2b7a.
